Android detection and packaging

tangalbert919 opened the pull request:
### Purpose of this PR

* Adds proper Android detection (no more proprocessing directives).
* Creates NuGet packages for Android.

### Testing status

* Testing currently in progress with an Android port of osu-framework to ensure the correct Factory is loaded. So far, detection is working.

### Comments

* If osuTK detects Android, `RunningOnLinux` is set to false to bypass SDL2 and X11 detection, and ensures we never load `LinuxFactory`. This does not affect desktop Linux machines.

tangalbert919 committed:
Add Android detection

tangalbert919 committed:
Make AndroidFactory and other classes public

tangalbert919 committed:
Create Android NuGet packages

VPeruS commented on src/osuTK.Android/AndroidFactory.cs:
```diff
@@ -8,10 +8,11 @@
 namespace osuTK.Android
 {
-    internal sealed class AndroidFactory : PlatformFactoryBase
+    public class AndroidFactory : PlatformFactoryBase
```
Why to remove sealed?

VPeruS commented on src/osuTK.Android/AndroidDisplayDeviceDriver.cs:
```diff
@@ -5,11 +5,12 @@
+namespace osuTK.Android
 {
-    internal class AndroidDisplayDeviceDriver : IDisplayDeviceDriver
+    public class AndroidDisplayDeviceDriver : IDisplayDeviceDriver
```
```suggestion
    public sealed class AndroidDisplayDeviceDriver : IDisplayDeviceDriver
```

VPeruS commented on src/osuTK.Android/GameViewBase.cs:
```diff
@@ -1139,6 +1139,8 @@ Icon INativeWindow.Icon
             set { throw new NotSupportedException(); }
         }
 
+        public bool CursorGrabbed { get => throw new NotImplementedException(); set => throw new NotImplementedException(); }
```
```suggestion
        public bool CursorGrabbed
        {
            get => throw new NotImplementedException();
            set => throw new NotImplementedException();
        }
```

tangalbert919 replied:
This seems unnecessary as it doesn't change the code.

VPeruS replied:
Similar code should use same formatting. Take a look at upper property

VPeruS commented on src/osuTK/Configuration.cs:
```diff
@@ -264,6 +258,23 @@ private static bool DetectX11()
+            Assembly[] assems = currentDomain.GetAssemblies();
+
+            //List the assemblies in the current application domain.
+            foreach (Assembly assem in assems)
```
```suggestion
            foreach (var assem in AppDomain.CurrentDomain.GetAssemblies())
```

VPeruS replied:
Also what about:
```csharp
var assembly = AppDomain.CurrentDomain.GetAssemblies().ToList().Find(a => a.ToString() == "Mono.Android");
return assembly != default;
```

VPeruS commented on src/osuTK/Platform/Factory.cs:
```diff
@@ -122,17 +122,18 @@ public Factory()
+                {
+                    Angle = new Egl.EglAnglePlatformFactory(Embedded);
+                }
+                
```
```suggestion

```

tangalbert919 replied:
This, I want to leave alone because even though Android doesn't need (or support) it, other platforms might.

VPeruS commented on src/osuTK/Platform/Factory.cs:
```diff
@@ -122,17 +122,18 @@ public Factory()
-#endif
+                if (!Configuration.RunningOnAndroid)
+                {
+                    Angle = new Egl.EglAnglePlatformFactory(Embedded);
```
Move this into upper else statement

VPeruS commented on src/osuTK/Configuration.cs:
```diff
@@ -273,12 +284,19 @@ internal static void Init(ToolkitOptions options)
 
-                    if (options.Backend == PlatformBackend.Default)
+                    // This will stop the LinuxFactory from loading on Android.
+                    if (RunningOnAndroid)
```
Check for RunningOnAndroid in upper if and prevent DetectUnix

VPeruS replied:
or add runningOnAndroid to DetectUnix

VPeruS commented on src/osuTK.Android/AndroidGraphicsContext.cs:
```diff
@@ -14,9 +14,10 @@
 
 namespace osuTK.Android {
-    internal class AndroidGraphicsContext : EglContext
+    public class AndroidGraphicsContext : EglContext
```
```suggestion
    public sealed class AndroidGraphicsContext : EglContext
```

swoolcock commented on src/osuTK.Android/GameViewBase.cs:
```diff
@@ -1139,6 +1139,8 @@ Icon INativeWindow.Icon
             set { throw new NotSupportedException(); }
         }
 
+        public bool CursorGrabbed { get => throw new NotImplementedException(); set => throw new NotImplementedException(); }
```
To be honest I would force this to always be true. We can’t assume that anything cursor-related is unsupported, since there are mice for tablets.

VPeruS replied:
Then you must guarantee that it always grabbed

swoolcock replied:
How are you going to click outside of a fullscreen application on Android? Unless you’re talking about overlays.  It’s not guaranteed on iOS unless you request always fullscreen in your info.plist, otherwise the user can interact with split screen apps.

qwewqa replied:
Android has multitasking as well and Chrome OS can display apps as windowed

swoolcock replied:
Regardless, I would rather a default functionality than an exception.  We can do a full implementation later, but I don't want to be explicitly avoiding calls to things that will throw on mobile.

tangalbert919 committed:
Take some (but not all) of @VPeruS's suggestions

Also target monoandroid50 if we want to get any NuGet packages built.

EDIT: Fix an ambiguity with `Size`

tangalbert919 committed:
Take some (but not all) of @VPeruS's suggestions

Also target monoandroid50 if we want to get any NuGet packages built.

tangalbert919 committed:
Merge branch 'android-packaging' of https://github.com/tangalbert919/opentk into android-packaging

peppy commented:
Could you please resolve the suggestions you applied above?

tangalbert919 commented:
Those applied suggestions have been resolved, followed by fixes made so appveyor can make the NuGet packages for Android successfully.

smoogipoo reviewed (approved):
Seems fine to me for now.
